File: packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FocusLockChannel.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// yuvrecorder patch: exposes a side method channel that locks the active
// AVCaptureDevice's focus at infinity via setFocusModeLocked(lensPosition: 1.0)
// (lensPosition 1.0 = farthest focal point = optical infinity).

import AVFoundation
import Flutter

#if canImport(camera_avfoundation_objc)
  import camera_avfoundation_objc
#endif

final class FocusLockChannel: NSObject {
  static let channelName = "dev.aircraft.yuvrecorder/camera_focus"

  private let channel: FlutterMethodChannel
  private weak var cameraPlugin: CameraPlugin?
  private let captureSessionQueue: DispatchQueue

  init(
    messenger: FlutterBinaryMessenger,
    cameraPlugin: CameraPlugin,
    captureSessionQueue: DispatchQueue
  ) {
    self.channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: messenger)
    self.cameraPlugin = cameraPlugin
    self.captureSessionQueue = captureSessionQueue
    super.init()
    self.channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }
  }

  func tearDown() {
    channel.setMethodCallHandler(nil)
  }

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "lockFocusAtInfinity":
      lockFocusAtInfinity(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func lockFocusAtInfinity(result: @escaping FlutterResult) {
    captureSessionQueue.async { [weak self] in
      guard let cam = self?.cameraPlugin?.camera as? FLTCam else {
        result(
          FlutterError(code: "no_camera", message: "Camera is not initialized yet.", details: nil))
        return
      }
      let device = cam.captureDevice.device
      guard device.isFocusModeSupported(.locked),
        device.isLockingFocusWithCustomLensPositionSupported
      else {
        result(
          FlutterError(
            code: "unsupported",
            message: "Device does not support locking focus at a custom lens position.",
            details: nil))
        return
      }
      do {
        try device.lockForConfiguration()
      } catch {
        result(FlutterError(code: "lock_failed", message: error.localizedDescription, details: nil))
        return
      }
      // lensPosition 1.0 = farthest = infinity.
      device.setFocusModeLocked(lensPosition: 1.0) { _ in
        device.unlockForConfiguration()
        result(nil)
      }
    }
  }
}

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/FocusLockChannel.java
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// yuvrecorder patch: exposes a side method channel that locks the active
// CameraX camera's focus at infinity by setting CONTROL_AF_MODE=OFF and
// LENS_FOCUS_DISTANCE=0 (0 diopters = infinity) via Camera2Interop.
//
// Re-applies the lock on every CameraX use-case rebind (e.g. startImageStream
// recreates the capture session and drops prior Camera2Interop options).

package io.flutter.plugins.camerax;

import android.hardware.camera2.CameraMetadata;
import android.hardware.camera2.CaptureRequest;
import android.os.Handler;
import android.os.Looper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.camera.camera2.interop.Camera2CameraControl;
import androidx.camera.camera2.interop.CaptureRequestOptions;
import androidx.camera.camera2.interop.ExperimentalCamera2Interop;
import androidx.camera.core.CameraControl;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

@ExperimentalCamera2Interop
final class FocusLockChannel implements MethodChannel.MethodCallHandler {
  static final String CHANNEL = "dev.aircraft.yuvrecorder/camera_focus";

  @Nullable private static volatile CameraControl activeCameraControl;
  // When true, every rebind that produces a fresh CameraControl re-applies
  // infinity focus automatically (CameraX drops Camera2Interop options on
  // use-case rebind, e.g. when startImageStream binds ImageAnalysis).
  private static volatile boolean reapplyAtInfinityOnRebind = false;

  static void setActiveCameraControl(@Nullable CameraControl control) {
    activeCameraControl = control;
    if (control != null && reapplyAtInfinityOnRebind) {
      try {
        Camera2CameraControl.from(control).addCaptureRequestOptions(infinityOptions());
      } catch (Throwable ignored) {
        // Best-effort; explicit method-channel calls surface real failures.
      }
    }
  }

  private static CaptureRequestOptions infinityOptions() {
    return new CaptureRequestOptions.Builder()
        .setCaptureRequestOption(
            CaptureRequest.CONTROL_AF_MODE, CameraMetadata.CONTROL_AF_MODE_OFF)
        .setCaptureRequestOption(CaptureRequest.LENS_FOCUS_DISTANCE, 0.0f)
        .build();
  }

  private final MethodChannel channel;
  private final Handler mainHandler = new Handler(Looper.getMainLooper());

  FocusLockChannel(@NonNull BinaryMessenger messenger) {
    channel = new MethodChannel(messenger, CHANNEL);
    channel.setMethodCallHandler(this);
  }

  void tearDown() {
    channel.setMethodCallHandler(null);
    activeCameraControl = null;
    reapplyAtInfinityOnRebind = false;
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    final CameraControl control = activeCameraControl;
    if (control == null) {
      result.error("no_camera", "Camera is not bound yet.", null);
      return;
    }
    switch (call.method) {
      case "lockFocusAtInfinity":
        reapplyAtInfinityOnRebind = true;
        applyOptions(control, infinityOptions(), result);
        return;
      default:
        result.notImplemented();
    }
  }

  private void applyOptions(
      @NonNull CameraControl control,
      @NonNull CaptureRequestOptions options,
      @NonNull MethodChannel.Result result) {
    try {
      Camera2CameraControl.from(control)
          .addCaptureRequestOptions(options)
          .addListener(
              () -> mainHandler.post(() -> result.success(null)),
              command -> command.run());
    } catch (Throwable t) {
      mainHandler.post(() -> result.error("focus_lock_failed", t.getMessage(), null));
    }
  }
}
